Client node adjustments

Apollon77 opened the pull request:
This PR is updating ClientNode with some convenient and practical enhancements.

* Adjust NetworkClient  from isDisabled to isEnabed including an option to define the startup behavior
* Add Peers.connect as main method to use to connect; enhance disconnect to also allow "unknown" peers
* Add "enableCluster" to dynamically add endpoints and behaviors for direct node usage
* add Peers.forAddress to add/get a ClientNode instance for a peer without commissioning
* adds filling known address into CommissioningClient state when a channel is estalished already (and a TODO touse that later)

The SubscriptionServer is not needing the node, it just needs to connect, so connect is used. OTA next will use "forAddresss"

lauckhart commented:
> The SubscriptionServer is not needing the node, it just needs to connect, so connect is used. OTA next will use "forAddresss"

Is this maybe just a case of "defaultSubscription = undefined"?

Apollon77 commented on packages/node/src/node/client/Peers.ts:
```diff
@@ -112,6 +119,41 @@ export class Peers extends EndpointContainer<ClientNode> {
+            const factory = this.owner.env.get(ClientNodeFactory);
+            node = factory.create({
+                ...options,
+                id: `${CLIENT_NODE_PREFIX}${NodeId.toHexString(peerAddress.nodeId)}`,
```
Thats exactly the discussion I mean above :-)
Yes you are right fabric id is missing in that here if we keep.

In my eyes the default naming of client node "ids" (a dev provided string or increasing number) is a bit intransparent because if makes things very hard to see. In my eyes the case that a controller dev uses "speaking name ids" (like the new Controller example) is rare because normal usecases are more automated/scripted. So most nodes will have just a number id (aka "epXXX") which is then again hard to see which real peer node it means.

In my ideal thinking we could have a numeric thing for anything not commissioned or such but then I think having the fabric+nodeid as identifier is clear in all places. But yes wanted to kickoff that discussion with this too :-)

Apollon77 replied:
Reverted to auto id

Apollon77 commented on packages/node/src/storage/client/ClientNodeStores.ts:
```diff
@@ -9,7 +9,11 @@ import type { ClientNode } from "#node/ClientNode.js";
 
-const CLIENT_ID_PREFIX = "peer";
+/**
+ * Prefix for incrementally assigned client IDs that are usually commissioned, so an increasing ID is used instead the
```
Yes see also above. I understood that, but unsure if the (in my eyes) downsides would be worth that, but open to discuss.

lauckhart commented on packages/node/src/behavior/system/network/NetworkClient.ts:
```diff
@@ -97,9 +111,18 @@ export namespace NetworkClient {
+         * If not defined the node will start up in the state that it was last in.
          */
-        isDisabled = false;
+        enabledOnStartUp?: boolean;
```
I was thinking that if you don't want the node to subscribe automatically you set startupSubscription to null, so "do not subscribe on startup"

Apollon77 replied:
For me a boolean flag is more convenient that always switching between "null" and "whatever subscription pattern the dev wants".

In fact the current reality is that if you do not enable the node you can not interact with it. maybe we could change that? Would allow node interactions a side (officially enable/disable aka main subscription"?

And as long as we need to enable the node I think controlling the behavior on startup is a good way.

lauckhart replied:
I made "isDisabled" after our conversation r.e. devs needing a mechanism to specify "the device is offline, don't use it".  So it's kind of just metadata anyway, would probably be OK to allow comms with disabled nodes.  The main downside is that there should be nothing to prevent unintended access to a node.

I assume the "startupSubscription" should still be ignored if the node is marked disabled?

lauckhart replied:
Separately, r.e. enabling the node on startup, I may be missing the use case.

Conceptually enableAtStartup sounds like "turn the device on when the controller restarts", but the controller doesn't actually know how to enable the node, just flip its local bit.

So what's the scenario where you mark the node as disabled but want it to change state when the controller restarts?

lauckhart commented:
> Did you mixed client and server now? The case her for subscription server is the re establishment of server side subscriptions. Yes when we somewhen do ServerInteraction then it might go there. But for now it goes into InteractionServer. 
> Or what did you mean?

I wasn't sure what you meant by "The SubscriptionServer is not needing the node, it just needs to connect, so connect is used."  Is this about a controller not allowing subscriptions then?

Apollon77 commented:
Alright @lauckhart, as discussed offline yesterday I adjusted the approach:
* isDisabled on Node networkClient can be used to disable a node, default is false. (aka enabled)
* The default subscription now also has this name and is always defined -  if not by the dev then it is a default full wildcard subscription. Optional, so undefined also means "default"
* autoSubscribe indicates if the node automatically subscribes on startup. Changing this leads to setup or teardown of the subscription
* special method "subscribe" on ClientNode allows to trigger a manual subscription after start if auto is disabled
* I added the logic for fabricFiltered Read checking from the other PR to here, so we can handle it together
* I moved the Initialization of ClientSubscriptions to be dynamic when peer nodes exist because in ControllerBehavior it was only initialized when this was added.

Apollon77 commented on packages/node/src/node/ClientNode.ts:
```diff
@@ -214,6 +219,80 @@ export class ClientNode extends Node<ClientNode.RootEndpoint> {
+     * The method adds the endpoint, if not existing, and also adds the cluster behavior for the "Complete" variant
+     * (means: all features, commands and attributes) of the cluster to the endpoint, if not already present.
+     */
+    async enableCluster(endpointId: number, type: ClusterBehavior.Type) {
```
done

Apollon77 commented on packages/protocol/src/action/Interactable.ts:
```diff
@@ -33,6 +33,11 @@ export interface Interactable<SessionT = InteractionSession> {
+    /**
+     * Cancel an active subscription.
+     */
+    cancelSubscription(id: number): void;
```
changed
